File: java/remoting/test/com/taobao/top/link/remoting/JsonSerializerTest.java
package com.taobao.top.link.remoting;

import static org.junit.Assert.*;

import java.nio.charset.Charset;
import java.util.Date;
import java.util.HashMap;

import org.junit.Test;

import com.taobao.top.link.LinkException;

public class JsonSerializerTest {
	@Test
	public void cross_methodCall_test() throws FormatterException {
		methodCall_test(new CrossLanguageJsonSerializer());
	}

	@Test
	public void cross_methodReturn_test() throws FormatterException {
		methodReturn_test(new CrossLanguageJsonSerializer());
	}

	private void methodCall_test(Serializer serializer) throws FormatterException {
		MethodCall call1 = new MethodCall();
		call1.MethodName = "echo";
		call1.TypeName = "serviceType";
		call1.Uri = "uri";
		call1.Args = new Object[] {
				"abc中文",// unicode support?
				(byte) 1,
				(double) 1.1,
				(float) 1.2,
				1,
				1L,
				(short) 1,
				new Date(),
				getMap(),
				getEntity(),
				new String[] { "abc" } };
		call1.MethodSignature = new Class<?>[] {
				String.class,
				byte.class,
				double.class,
				float.class,
				int.class,
				long.class,
				short.class,
				Date.class,
				HashMap.class,
				Entity.class,
				String[].class };

		byte[] ret = serializer.serializeMethodCall(call1);
		System.out.println(new String(ret, Charset.forName("UTF-8")));
		MethodCall call2 = serializer.deserializeMethodCall(ret);
		assertEquals(call1.MethodName, call2.MethodName);
		assertEquals(call1.TypeName, call2.TypeName);
		assertEquals(call1.Uri, call2.Uri);
		assertEquals(call1.Args.length, call2.Args.length);
		for (int i = 0; i < call1.Args.length; i++) {
			assertEquals(call1.Args[i].getClass(), call2.Args[i].getClass());
			assertEquals(call1.MethodSignature[i], call2.MethodSignature[i]);
		}
		for (Object arg : call2.Args)
			System.out.println(String.format("%s|%s", arg.getClass(), arg));
	}

	private void methodReturn_test(Serializer serializer) throws FormatterException {
		MethodReturn _return1 = new MethodReturn();
		_return1.Exception = new LinkException("error", new NullPointerException());
		_return1.ReturnValue = getEntity();

		byte[] ret = serializer.serializeMethodReturn(_return1);
		System.out.println(new String(ret, Charset.forName("UTF-8")));

		MethodReturn _return2 = serializer.deserializeMethodReturn(ret, Entity.class);

		System.err.println(_return2.Exception.getMessage());
		_return2.Exception.printStackTrace();

		assertEquals(_return1.ReturnValue.getClass(), _return2.ReturnValue.getClass());
		assertEquals(((Entity) _return1.ReturnValue).getString(), ((Entity) _return2.ReturnValue).getString());
		System.out.println(((Entity) _return2.ReturnValue).getMap());
		assertEquals(((Entity) _return1.ReturnValue).getMap(), ((Entity) _return2.ReturnValue).getMap());
		assertEquals(((Entity) _return1.ReturnValue).getMap().size(), ((Entity) _return2.ReturnValue).getMap().size());
		assertEquals(((Entity) _return1.ReturnValue).getArray()[0], ((Entity) _return2.ReturnValue).getArray()[0]);
	}

	private HashMap<String, String> getMap() {
		HashMap<String, String> map = new HashMap<String, String>();
		map.put("k", "k");
		return map;
	}

	private Entity getEntity() {
		Entity e = new Entity();
		e.setString("string");
		e.setMap(getMap());
		e.setArray(new String[] { "abc" });
		return e;
	}
}

File: csharp/src/Top.Link.Test/JsonSerializerTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using NUnit.Framework;
using Taobao.Top.Link.Remoting;
using Taobao.Top.Link.Remoting.Serialization;
using Taobao.Top.Link.Remoting.Serialization.Json;

namespace Taobao.Top.Link.Test
{
    [TestFixture]
    public class JsonSerializerTest
    {
        private static ISerializer jsonSerializer = new CrossLanguageJsonSerializer();

        [TestCase]
        public void TypeTest()
        {
            Console.WriteLine(typeof(string[]));
            Console.WriteLine(typeof(Entity[]));
            var now = DateTime.Now;
            Assert.AreEqual(now, new DateTime(now.Ticks));
        }

        [TestCase]
        public void MethodCallTest()
        {
            MethodCall call1 = new MethodCall();
            call1.MethodName = "echo";
            call1.TypeName = "serviceType";
            call1.Uri = "uri";
            call1.Args = new Object[] {
		        "abc中文",
		        (byte) 1,
		        (double) 1.1,
		        (float) 1.2,
		        1,
		        1L,
		        (short) 1,
		        DateTime.Now,
		        GetDictionary(),
		        GetEntity(),
		        new string[] { "abc" } 
            };
            call1.MethodSignature = new Type[] {
		        typeof(string),
		        typeof(byte),
		        typeof(double),
		        typeof(float),
		        typeof(int),
		        typeof(long),
		        typeof(short),
		        typeof(DateTime),
		        typeof(IDictionary<string,string>),
		        typeof(Entity),
		        typeof(string[])
            };

            byte[] ret = jsonSerializer.SerializeMethodCall(call1);
            Console.WriteLine(Encoding.UTF8.GetString(ret));

            MethodCall call2 = jsonSerializer.DeserializeMethodCall(ret);
            Assert.AreEqual(call1.MethodName, call2.MethodName);
            Assert.AreEqual(call1.TypeName, call2.TypeName);
            Assert.AreEqual(call1.Uri, call2.Uri);
            Assert.AreEqual(call1.Args.Length, call2.Args.Length);
            for (int i = 0; i < call1.Args.Length; i++)
            {
                Assert.AreEqual(call1.Args[i].GetType(), call2.Args[i].GetType());
                Assert.AreEqual(call1.MethodSignature[i], call2.MethodSignature[i]);
                Assert.AreEqual(call1.Args[i].ToString(), call2.Args[i].ToString());
            }
            foreach (object arg in call2.Args)
                Console.WriteLine(string.Format("{0}|{1}", arg.GetType(), arg));
        }

        [TestCase]
        public void MethodReturnTest()
        {
            MethodReturn return1 = new MethodReturn();
            return1.Exception = new LinkException("error", new NullReferenceException());
            return1.ReturnValue = GetEntity();

            byte[] ret = jsonSerializer.SerializeMethodReturn(return1);
            Console.WriteLine(Encoding.UTF8.GetString(ret));

            MethodReturn return2 = jsonSerializer.DeserializeMethodReturn(ret, typeof(Entity));
            Console.WriteLine(return2.Exception.Message);
            Assert.AreEqual(return1.Exception.Message, return2.Exception.Message);
            Assert.AreEqual(return1.ReturnValue.GetType(), return2.ReturnValue.GetType());
            Assert.AreEqual(((Entity)return1.ReturnValue).String, ((Entity)return2.ReturnValue).String);
            Assert.AreEqual(((Entity)return1.ReturnValue).Dictionary, ((Entity)return2.ReturnValue).Dictionary);
            Assert.AreEqual(((Entity)return1.ReturnValue).Dictionary.Count, ((Entity)return2.ReturnValue).Dictionary.Count);
            Assert.AreEqual(((Entity)return1.ReturnValue).Array[0], ((Entity)return2.ReturnValue).Array[0]);
        }

        private IDictionary<String, String> GetDictionary()
        {
            IDictionary<string, string> map = new Dictionary<string, string>();
            map.Add("k", "k");
            return map;
        }
        private Entity GetEntity()
        {
            Entity e = new Entity();
            e.String = "string";
            e.Dictionary = this.GetDictionary();
            e.Array = new String[] { "abc" };
            return e;
        }

        public class Entity
        {
            public string String { get; set; }
            public IDictionary<String, String> Dictionary { get; set; }
            public string[] Array { get; set; }
        }
    }
}